Assets/MixedRealityToolkit.Services/InputSystem/FocusProvider.cs

```diff
@@ -854,70 +840,30 @@ private bool RaycastGraphicsStep(PointerData pointerData, RayStep step, LayerMas
         /// </summary>
         private void UpdateFocusedObjects()
         {
-            Debug.Assert(pendingPointerSpecificFocusChange.Count == 0);
-            Debug.Assert(pendingOverallFocusExitSet.Count == 0);
-            Debug.Assert(pendingOverallFocusEnterSet.Count == 0);
-
-            // NOTE: We compute the set of events to send before sending the first event
-            //       just in case someone responds to the event by adding/removing a
-            //       pointer which would change the structures we're iterating over.
-
+            // Collect all pointers that have their target changed:
             foreach (var pointer in pointers)
             {
-                if (pointer.PreviousPointerTarget != pointer.CurrentPointerTarget)
-                {
-                    pendingPointerSpecificFocusChange.Add(pointer);
+                IMixedRealityPointer currentPointer = pointer.Pointer;
+                GameObject pendingUnfocusObject = pointer.PreviousPointerTarget;
+                GameObject pendingFocusObject = pointer.CurrentPointerTarget;
 
-                    // Initially, we assume all pointer-specific focus changes will
-                    // also result in an overall focus change...
+                if (pendingUnfocusObject != pendingFocusObject)
+                {
+                    MixedRealityToolkit.InputSystem.RaisePreFocusChanged(currentPointer, pendingUnfocusObject, pendingFocusObject);
 
-                    if (pointer.PreviousPointerTarget != null)
+                    if (pendingUnfocusObject != null)
                     {
-                        pendingOverallFocusExitSet.Add(pointer.PreviousPointerTarget);
+                        MixedRealityToolkit.InputSystem.RaiseFocusExit(currentPointer, pendingUnfocusObject);
                     }
 
-                    if (pointer.CurrentPointerTarget != null)
+                    if (pendingFocusObject != null)
                     {
-                        pendingOverallFocusEnterSet.Add(pointer.CurrentPointerTarget);
+                        MixedRealityToolkit.InputSystem.RaiseFocusEnter(currentPointer, pendingFocusObject);
```

Comment:
>MixedRealityToolkit.InputSystem.RaiseFocusEnter(currentPointer, pendingFocusObject); [](start = 24, length = 84)

Raising focus enter/left for all pointers will break all focus handlers relying on the existing behavior. We already have focus changed events to detect multiple pointers,